### DISCUSSION TITLE
General chat

### DISCUSSION THREAD

**Discussion by silverwings15** [aminomancer/uc.css.js#4] [General]
hey MS, it's mime from reddit 👋 looks like you've been putting some massive work into the proton stuff. i'm still waiting for v89 to hit beta channel before i try it out, so a little more than a month left to go. just checking in on you and saying thanks for the Ctrl W closes pinned tab script, brilliant stuff really. such a simple idea yet a massive improvement over default behavior for me.

**Comment by aminomancer** [comment]
thanks man sorry I took so long to reply, I didn't realize anyone was using this tab and it sends the email updates to my spam inbox. I still haven't worked out a light mode, been pretty busy just keeping up with the proton changes, but I have converted a lot more color values and other stuff to variables in the process so making a light mode should be easier now

**Reply by silverwings15** [reply]
no worries. seems like everyone is freaking out over the proton changes, hope your fixes will alleviate the issues

**Reply by aminomancer** [reply]
I kept some of the changes but undid a lot more. some of them are irreversible, like the changes to the tab markup... but if they end up removing the tab sound icon completely I'm gonna try recreating it with a script. should be possible, although it's a huge nightmare since I don't think custom elements can be overridden. couple weeks ago I tried overriding the definition of "tabbrowser-tab" before any windows are loaded, and it worked, but it made the UI look totally deformed and messed up lol. not really sure why, seemed like I did everything right. so if they do end up removing it, I think the only method I'll have to recreate it is to listen for the creation/restoration of tabs, and add sound icons directly that way. which is way slower than the previous method I tried. so maybe I'll just be forced into accepting the new paradigm.

that's kind of a special case though, everything else I'm trying to keep stable. like the icons, now they're changing the bloody icons. maybe I'm crazy but I like the old icon style, it's unique and bold and even kinda charming. these thick icons are perfect for animation too. the new ones they're replacing everything with look really corny and derivative and honestly amateurish. a few of the old ones could use some work but I already put too much time into redesigning them myself to back down now. so I guess this theme will directly control icons from now on and I'll be spending a lot more time making new ones.

**Reply by silverwings15** [reply]
i see you adding proton fixes even now, are they still rolling out changes on a regular basis?

**Reply by aminomancer** [reply]
yeah it hasn't slowed down at all. huge changes to the icons lately. it was quite a nightmare, I decided to completely change the way I deal with icons. not just because there are so many elements to manually change the icons for, but also because some things like favicons can't be overridden with CSS. so now instead of setting CSS rules to change icon properties like list-style-image etc., I use this manifest syntax to directly override the chrome URLs. so I can remap chrome://browser/skin/whatever.svg to whatever I want, which changes every instance across the whole browser. with fx-autoconfig you can just add rules to the existing chrome.manifest file, and they are soooo much faster to type. though data URIs don't work to my knowledge so you have to keep the actual icon files locally somewhere in the chrome folder.

you can see the syntax in the utils/chrome.manifest file on here. just like `override chrome://blahblah ../resources/whatever.svg`. the only real nuance to it is `./` means the current folder (the folder where chrome.manifest is located) and `../` means the next folder up. so since chrome.manifest is in utils folder, and both resources and utils are in chrome folder, we gotta use `../resources/`. if you had a file in the actual chrome folder you'd just use `../file.svg` and if you had to go up more than one folder, I assume you can use 3 dots or more.

other than that, it seems like the parent UI has been pretty stable, they've recently been changing a lot of stuff that affects UI content documents like the about: pages and dialogs.

**Reply by silverwings15** [reply]
so mozillas plan is to have all the proton stuff finalized by 89 stable right? or will they keep expanding on it in each new version?

wow i must try this manifest syntax you mention, i used to just mod omni.ja but it got too tedious after each update

**Reply by aminomancer** [reply]
I don't think it'll be finalized any time soon, it's just enabled by default in 89. Like, 89 stable isn't very different from 89 nightly, the stable release is just several releases behind the nightly release. They're on 90 in nightly and stuff is still changing. It won't get added retroactively to 89 since everything needs time to be tested and such. So whatever comes in stable's 89 will not be exactly the same as what's in 90 right now. But I expect the stable release will have all the proton stuff enabled by default, and maybe it'll have all the new icons.

yeah no need to modify the package when you can just override the manifest. I haven't tried it with anything but icons yet but I assume you can do this with anything. like, you could modify a lot of the source code that way I think. I don't think any of the C++ or rust stuff is loaded by URI, just the js/html/css stuff. but a lot of that stuff is hard to interact with directly otherwise so it could be pretty useful. like if you wanted to change some really fundamental stuff like ua.css, or a custom element's markup template, I think you could do it by overriding the URI with your own version. if they remove the tab sound icon I might do that to change tabbrowser-tab's markup and class methods, since there doesn't seem to be any reasonable way to do it with an ordinary script.

**Reply by silverwings15** [reply]
was gonna DL the 89 beta for a proton test run but now i have cold feet 😬. would be pretty sick if you can get the light theme ready in a few months when you're done with all this proton modification, which honestly sounds like quite the PITA to work with. also, do you know if it's possible to customize the findbar result indicator/tick? i find them too thin and not at all visible enough, and blue is an okay color if slightly boring 😁

**Reply by aminomancer** [reply]
findbar result indicator? can you be more specific what you're referring to? I haven't seen the vanilla findbar in like 3 years. my findbar is really heavily modified, both visually and DOM-wise. honestly can't remember what the normal one looks like. if what you're talking about is the results count ("1 of 200 matches" or whatever it says) I have a script that completely replaces it with a mini indicator placed in the findbar textbox area. could be modified pretty easily I imagine. or do you mean the next/previous arrow buttons?

**Reply by silverwings15** [reply]
hmm i'm surprised you haven't noticed it, it was introduced in 87. basically it shows a little tick for each search result at the edge of the screen where the scrollbar is
![image](https://user-images.githubusercontent.com/10742312/115482753-92a59600-a279-11eb-84f9-3010e3a43d34.png)

the same ticks in Chrome for example are much thicker and thus way more visible 
![image](https://user-images.githubusercontent.com/10742312/115482891-e617e400-a279-11eb-90c8-a459bc653b79.png)

**Reply by aminomancer** [reply]
oh, yeah I do know what you're talking about. they're called [slider marks](https://searchfox.org/mozilla-central/source/layout/xul/nsSliderFrame.cpp#233). yeah I know what you mean about the visibility. I'd like to be able to make them brighter and thicker too. I assumed you were talking about something in the actual findbar container. no, I don't think you can do anything to change them, except maybe it's possible to set ui.textHighlightBackground to an 8-digit hex code with alpha 0. in which case they'd be transparent but so would other highlights generated by that C++ LookAndFeel module. including the regular findbar highlighting. not a great solution, and I'm not even sure if it'll accept an 8-digit hex code in the first place. but I asked the devs on your behalf [here](https://bugzilla.mozilla.org/show_bug.cgi?id=259640), you should back me up so we look like a major interest group lol.

**Reply by silverwings15** [reply]
haha i didn't know that's the official name, Moz used the term tick in the changelog during rollout so i just went with that. i also didn't make the connection that it uses the same color as the findbar hits, that makes sense. thanks for delivering the message, i will also create a bugzilla account right away lol

**Reply by aminomancer** [reply]
Hey I think it was you who asked about the collapsing overlay scrollbars? Specifically about how they seem to be invisible sometimes when the page loads, and don't appear until you move your mouse over them? Well it turns out this is due to a pref, `layout.css.cached-scrollbar-styles.enabled`, it just needs to be set to false and the problem apparently goes away

**Reply by silverwings15** [reply]
wow nice find, i'll set it to false and test for a while. i feel like it was already happening a lot less since i updated to v88 (placebo?) so the effect won't be as noticeable, but nevertheless am happy to have the issue fully stamped out

**Reply by aminomancer** [reply]
hmm yeah maybe. I never noticed it happening inside an actual browser window, but I noticed it would happen like 99% of the time when opening a panel in the UI. like for the history toolbar button or all tabs button, when their popups had a scrollbar, it would almost always be invisible until I moused over its general location. I always figured that was some kind of value caching. I think it may also be the reason I ran into problems with my old script that was designed to expand the scrollbar while mousewheel scrolling. I wanted to set it up so that it would expand when hovering it, but also when just scrolling with the mousewheel. but every time I tried it, it'd expand but it'd get stuck in its expanded state until mousing over it. seems like a very similar pattern, which I guess is just firefox caching a whole bunch of CSS values because it doesn't expect them to change — they're normally set in stone. so I'm gonna try reevaluating that soon.

and yeah, it's a very lucky find, as it happens I was just experimenting with some of the weird look and feel [scrollbar prefs](https://searchfox.org/mozilla-central/source/widget/LookAndFeel.h#43) that aren't publicly documented or discussed anywhere, when it occurred to me I don't think I ever bothered to look at the basic scrollbar stylesheet. and then I saw [this](https://searchfox.org/mozilla-central/source/toolkit/content/minimal-xul.css#35) which just blew my mind and it suddenly it all made sense lol.

**Reply by silverwings15** [reply]
haha that's quite fortunate that you'd stumble on it like that

**Comment by silverwings15** [comment]
your repo is seriously amazing now - i tested everything on a new profile the other week and was seriously impressed. i tried adding some of the scripts in my current profile to it though and it didn't work at all - guess i need to do more than just dropping them into the scripts folder, but i haven't got to figuring that out yet. 

also, remember when you said you wanted ideas for scripts a few months back? i just thought of one. you know how there are addons that let you jump back and forth between your most recent tabs by clicking on the tab itself? i always found them quite annoying to use because sometimes you'd just click on the tab unintentionally and you jump off to a tab that's really far away, and since the new tab isn't focused on (at least not on TST), you'd have to scroll quite a bit to search for it so you can click it to get back to the tab you were originally on. however, what if you can accomplish the same thing via a key combo instead? that would eliminate the whole misclicking issue completely

leme know what you think!

**Reply by aminomancer** [reply]
hmm, which scripts didn't work? I've made some changes to how this stuff fundamentally works, so maybe you need to reinstall fx-autoconfig and other stuff. if you're also using the theme you would also need to replace the resources folder and `utils/chrome.manifest`. and maybe you need to set some prefs. prefs shouldn't be required for the scripts to work, but I guess you could try downloading the user.js file just in case.

as for a key combo that switches between recently used tabs, you can just set `browser.ctrlTab.sortByRecentlyUsed` to true and it already does that. or is your idea different? I can't say I've heard of addons that do something like that so idk if it's the same thing. I don't browse addons very often anymore. maybe you can send an example?

edit: I guess ctrl+tab doesn't work on macOS since there's already a system hotkey. seems weird that Firefox doesn't remap it to alt+tab or something on macOS. so if you're using macOS then maybe it would be better to just remap the ctrl+tab system. that might be difficult since it's not exactly a typical hotkey, it's a pretty elaborate system. but it's certainly better than making a new hotkey that can only switch to the most recent tab.

**Reply by silverwings15** [reply]
wow, all these years and i never knew about Ctrl-Tab.. exactly what i want. so glad you showed me this

the scripts i mentioned are just some of Aris and xiaoxiaoflood's scripts that i'm currently using and wanted to transfer over to your setup by dropping them into the 'script' folder. do i have to add their names to one of the .css files first? also, i just found out that i was using an older version of fx-autoconfig, so i'll make a fresh installation on the weekend and test everything again

**Reply by aminomancer** [reply]
no, CSS doesn't have anything to do with scripts. I mean some scripts might work in tandem with CSS but that's not a general thing. most likely the scripts are just incompatible with fx-autoconfig. updating fx-autoconfig won't change that since it works in a very different way. the updates just relate to a new utility function that can refresh user stylesheets. I don't know of any scripts that actually use it yet.

most of the scripts for xiaoxiaoflood's loader definitely won't work in fx-autoconfig without significant changes. that's why I have ports of some of those scripts on here, the originals only work with xiaoxiaoflood's loader, for several reasons. not small or easy things to change either. that loader runs scripts before any windows have opened, and then executes some specified function at startup. fx-autoconfig can do that with an optional directive, but in that mode it doesn't have access to any global objects provided by the loader. the xiaoxiaoflood scripts generally rely heavily on this global `_UC` object. fx-autoconfig has something similar in `_ucUtils` but it's not global, it's per-window and it's just for utilities. there's a function that can instantiate global objects which can be retrieved from different windows but that too requires significant changes to the xiaoxiaoflood scripts. I don't know about Aris though, I'm not sure what that is.

**Comment by silverwings15** [comment]
Aris is another guy who does FF customization, you can find his repos here:

https://github.com/Aris-t2/CustomJSforFx/

https://github.com/Aris-t2/CustomCSSforFx

the xiaoxiaoflood scripts that i'm using right now all work with fx-autoconfig, or at least the version of it from last year that i currently have, so i feel like compatibility shouldn't be an issue here. will investigate further and let you know

**Comment by aminomancer** [comment]
Well if you can send me the specific scripts that aren't working I can definitely tell you why. Excluding that, it's anyone's guess. The scripts on the Aris-t2 repo look like they should work fine with fx-autoconfig actually. At least the ones I checked. They should work, though they declare a lot of unscoped variables with the `var` keyword instead of `let` or `const` which isn't great. Could end up redeclaring existing global variables, since some of firefox's older JS components do still use `var` outside of blocks or modules. Or it could at least get them confused within the script's context.

Aside from compatibility with old browsers (which isn't a factor in what we're doing obviously) there isn't much reason to use `var` anymore unless you specifically are trying to make the variable unscoped. But there are safer ways of doing that, like properties. That's why I use mostly classes now, it's just the minimal intrusion on Firefox's built-in scripts. So if I was gonna use some of these I'd probably want to refactor them slightly. It's easy enough to replace all the `var` declarations with `let` though. There's other stuff too, like wrapping the script in an IIFE and waiting for delayed startup but that's more complicated. So I'd suggest just send me whatever script you can't get to work, these all seem to be pretty short scripts that can be adjusted quickly.

**Comment by silverwings15** [comment]
aside from some of your scripts that are already part of this repo, i'm using the following:

https://github.com/Aris-t2/CustomJSforFx/blob/master/scripts/editbookmarkspopup_expanded.uc.js
https://github.com/Aris-t2/CustomJSforFx/blob/master/scripts/favicon_in_urlbar.uc.js
https://github.com/Aris-t2/CustomJSforFx/blob/master/scripts/restart_button.uc.js
https://github.com/Aris-t2/CustomJSforFx/blob/master/scripts/addonbar.uc.js
https://github.com/xiaoxiaoflood/firefox-scripts/blob/master/chrome/rebuild_userChrome.uc.js
https://github.com/mimecry/dotfiles/blob/master/persistent_mininal_scrollbar.uc.js (your scrollbar script that you shared some time back which i believe is not part of this repo)

would be awesome to these integrated into your setup, especially the rebuild_userchrome one as it allows quick and simple disabling/reenabling of scripts

**Comment by aminomancer** [comment]
The scrollbar is still in this repo, it's just in `userChrome.ag.css`, which gets loaded by `userChromeAgentAuthorSheetLoader.uc.js`. Just a cleaner way of doing things that don't require dedicated javascript. So you don't need that script anymore. You just need the agent sheet and the loader, plus a pref `layout.css.cached-scrollbar-styles.enabled` which needs to be false.

The rebuild userChrome script can't work in fx-autoconfig FYI, it's just a fundamental difference between the loaders. fx-autoconfig always requires a restart when scripts are disabled/enabled. Even if it had that feature, it wouldn't work because it all depends on a script. Only reason that feature works is because xiaoxiaoflood puts special behavior for initialization and destruction in every script. None of my scripts have anything like that, so they wouldn't know what to do to disable themselves, and they also rely on startup queues that won't exist when they're re-enabled dynamically. Some of them can *only* run during startup. It's a limitation but I don't ever disable scripts anyway, if I don't wanna use a script then I just delete it.

I don't see any reason why `editbookmarkspopup_expanded.uc.js` wouldn't work in fx-autoconfig. Are you sure it works with any loader? It might just be old. Same with the other Aris scripts. They are designed to be loaded by a similar loader to fx-autoconfig. A bit simpler but basically the same fundamental method. So what works for that loader should work for fx-autoconfig. You can see just looking at them what a big difference there is between xiaoxiaoflood's scripts vs mine and Aris-t2's and Alice0775's. There's the window method, which all ours use, and there's the global method, which most of xiaoxiaoflood's scripts use. So I expect all of Alice0775's and Aris-t2's scripts to be compatible with fx-autoconfig.

**Reply by silverwings15** [reply]
so i figured out that it isn't the scripts i'm adding that aren't working, it's that none of the scripts in the repo work at all. i went back to read the instructions and noticed that you mention it's only confirmed working for Nightly now, so i'll work on getting that set up and then report back

**Reply by aminomancer** [reply]
Eh, the scripts do work in stable, it's just the theme that might look weird in stable

**Reply by aminomancer** [reply]
you sure you aren't missing something? do you get the autoconfig startup warning in the console?

**Comment by silverwings15** [comment]
wow i still can't get your scripts working no matter what i try. this appears on startup so fx-autoconfig seems to be working properly 
![image](https://user-images.githubusercontent.com/10742312/127177371-8799c621-9957-4bcc-a846-e1eead575bad.png)

here is how i added your files, i'm quite sure they're all in the correct locations. this is driving me nuts lol

![image](https://user-images.githubusercontent.com/10742312/127178304-bcc9e4fc-5759-4db4-a233-c820a3afb6f8.png)

![image](https://user-images.githubusercontent.com/10742312/127177977-007883e7-d3dc-4f80-b6ab-caa8388e3fc6.png)

![image](https://user-images.githubusercontent.com/10742312/127178023-ed0df929-83f0-4160-98a8-6bbd5cdb7f40.png)

![image](https://user-images.githubusercontent.com/10742312/127178064-d7bee0c8-4332-4b72-94d4-e95d09578aa1.png)

**Reply by aminomancer** [reply]
show me what chrome.manifest says

**Comment by aminomancer** [comment]
show me what chrome.manifest says

**Comment by aminomancer** [comment]
you sure chrome.manifest is searching in the `script` folder and not the `JS` folder? you might need to update the paths in the manifest to match the paths in your folder.

also, maybe it's due to the name of the profile folder. I doubt it, but you should check to make sure the name of the profile folder actually matches the name of the profile in profiles.ini and installs.ini. in the firefox roaming appdata folder. and maybe try making a new profile with a more generic name, let firefox name it that is.

**Comment by aminomancer** [comment]
if all else fails, try replacing all the content of boot.jsm [with this](https://paste.in/aVpQv7), and replacing all the content of config.js [with this](https://paste.in/zqmvCK)

**Comment by silverwings15** [comment]
oh man you got it, i pulled chrome.manifest from the fx-autoconfig repo and didn't notice he'd made that particular change. playing around with all your scripts now and they're so beautiful...

**Reply by aminomancer** [reply]
oh, in that case you definitely need to get the full chrome.manifest file from my repo then. it's not just the script vs JS folder thing, mine has all the icon and script replacements. whereas his only has 3 lines, just the ones needed for the basic functionality of the script loader.

**Reply by silverwings15** [reply]
yep i went ahead and did that when i realized the icons were not properly displaying. this is absolutely amazing stuff, you're hands down the best 'content creator' as far as firefox customization is concerned at the moment

**Reply by aminomancer** [reply]
haha thanks, never thought about it that way before. I guess it is content. I have to get back into it, I've been pretty strapped for javascript ideas lately. and I don't think my CSS stuff is very popular, it seems like everyone is obsessed with "one-line" toolbar layouts recently. but I just don't see a way to make that look good without reverting tabs back to proton's button-like appearance. so the repo has been kinda stagnating these past weeks

**Reply by silverwings15** [reply]
i'm also in the camp of 'more space = more room to display content' so i can defo understand why. even though it is sacrilegious to say here :D, as a TST user, i don't think i can ever go back to a setup that utilizes a horizontal tab bar. that and the color theme aside, i would defo steal your CSS in a heartbeat

**Reply by aminomancer** [reply]
yeah I figured I'd try making a vertical tab layout since it seems so popular now. put an enormous amount of effort into doing it with autoconfig too, since tab switcher addons look and feel so sketchy to me. especially their context menus. and of course they basically hijack the sidebar feature. after I made it, I tried hiding the horizontal tab bar, since what use is a vertical tab bar if it's not conserving vertical space. but the navbar just felt so barren and unnaturally small. maybe I'm just so used to this layout I can't shake it, but I feel like the overall screen composition, the proportions objectively look off like that. it'd be like if I made the navbar only 10px tall on a 1080p window. I'm an old graphics guy, I guess I have a lot of strong opinions about what constitutes good design.

so in the end although I have it installed, I almost never use my own vertical tabs script lol. I mostly turn it on just to admire it and pat myself on the proverbial back. it's not quite up to par with TST anyway, I guess the main feature of TST is the "tree" part. visible and modifiable tab hierarchies. it's funny, firefox makes it easier for addons to do that than for itself. addons are worse in almost every way and almost never look like a real part of the UI, but for whatever reason the tabs API exposes methods and properties that don't really exist in the tab browser itself. it's still possible to make that happen but it's a lot of time to spend on something I don't even use myself

**Reply by silverwings15** [reply]
haha yeah the vertical pane looks nice and is super smooth and responsive (especially the tab dragging, fockin love that), but features wise it just can't compete with TST and its army of helper addons.

i know you said not to attempt making a light theme right now, but i'm still gonna try and play around with the hue values, see what happens

**Reply by aminomancer** [reply]
it should be a lot easier now than it was in the past. almost everything important is integrated with uc-globals.css and uc-low-globals.css. you will have to deal with some of the `prefers-color-scheme` media queries though. I made those when I was attempting to make a light mode, but in many cases I never added an actual light version. so the rules in those queries don't have any light mode corollary. so you'll have to duplicate them and put them outside and above the media query. it's good though, I don't mind at all and if you come up with anything feel free to make a PR or upload it some other way.

**Comment by silverwings15** [comment]
so i have this idea that you perhaps can try working into your findbar script. a while back i think i told you about this small bit of code i was using, that hides my floating findbar whenever it goes out of focus

```
/* Findbar disappears when unfocused */
findbar:not(:focus-within) {
  height: 0 !important;
  overflow: hidden !important;
  transition: height 0.0s ease 0s !important;
}
```
however, since version 89 it no longer functions correctly. instead of hiding the bar, it just collapses it but not completely, like this 

![image](https://user-images.githubusercontent.com/10742312/129419551-45915585-1c5d-498d-a550-ec93f78c17c1.png)

i asked Aris if there was a way to fix it via CSS, and he said 

> I'm surprised there even was a way to achieve this. The findbar node does not offer any attributes to work with.

so am wondering if it is at all possible to restore this functionality via JS. what do you think?

**Reply by aminomancer** [reply]
Not sure why javascript would be necessary. `:focus-within` is all you need, if you used javascript to add an attribute associated with focus state, it would just be functionally equivalent.

And that wouldn't solve the issue anyway. I mean, your selector is clearly working, so you don't need javascript to capture the state. And the properties are applying too. You just don't have all the right properties you need to fully shrink the findbar, I guess because since proton, padding was added to the findbar.

What's the rest of the CSS? I don't know if my findbar is styled quite like yours, can't really tell from the screenshot. I tested the ruleset with my theme and it did work, there was just padding in the findbar spreading its bounding box. So this worked for me

```
findbar:not(:focus-within) {
    overflow: hidden !important;
    height: 0 !important;
    padding-block: 0 !important;
    transition: height 0.0s ease 0s !important;
}
```

And idk, maybe I'm misunderstanding something, but it's not surprising at all that it works. That's like saying you need to set an attribute on mouseover/mouseleave to convey hover state to CSS when there's already a :hover pseudo-class.

**Reply by silverwings15** [reply]
wow, spot on. your mastery of firefox modding is evident, even compared to other knowledgeable people 👍. thanks man

**Reply by aminomancer** [reply]
Haha oh thanks, so that worked?

**Reply by silverwings15** [reply]
yep worked perfectly :)

**Comment by silverwings15** [comment]
had a bit of time to play around with your latest findbar CSS along with the autohide CSS that i asked about previously.. one thing i can't replicate is to get it to respect your current animation settings. i played around with both `transition` and `transition-duration` values, but the fade in/fade out animation doesn't trigger no matter what. any idea about this one?

**Reply by aminomancer** [reply]
That's because you're using properties that are not animatable at all, like `overflow`. Technically `height` is animatable but only under certain circumstances, and I don't know what those circumstances are. I have often had problems animating it with some display modes. So I usually just don't use it at all. I also think it looks better to move the element than to squish it down.

But if you were going to animate its squishing down, you don't want to use one like `height` because that causes a reflow. Like, all the child elements' positions and size need to be recalculated formally because their container's height has changed. If you imagine changing the `height` of just a `<p>` element full of text, you can imagine how the flow of the text needs to change in accordance somehow. Or it needs to overflow off the element. So that's a reflow, a recalculation of the positions and sizes of elements. Whereas if you used `transform: scaleY(50%)` or something, you could get the same effect but it would not reflow.

Idk the technical details of how that works, but at the very least it's *as if* a frame buffer was taken and the frame buffer was squished down. So the positions and orientations of the child elements don't change. They all remain the same in relation to each other, only the proportional size of the whole thing changes. As if you took a screenshot of the findbar and then squished its height down. So that's generally gonna be a lot better for most animations.

I did that in the urlbar results, albeit very very slightly. Scale animations on big, complex objects with lots of children are often kind of ugly if they're too extreme. So that one's scale only changes to like 99.99% or something. But in this case the object is small enough I don't think that effect would look good or even be noticeable. So I just use `transform: translateY(-2em)` which, paired with `opacity: 0` is good enough.

I figured your use of `height` and `overflow` as the means of hiding the findbar meant you wanted it to have no animations. If you're using my stylesheet uc-findbar.css then you should just get rid of the ruleset you originally posted, with the `height` and `overflow` rules, and replace the fourth ruleset in uc-findbar.css...
```
findbar[hidden] {
    margin-bottom: auto !important;
    visibility: visible !important;
    transition: all 0.2s ease-in-out 0s !important;
    pointer-events: none;
    transform: translateY(-2em);
}
```
with this:

```
findbar[hidden],
findbar:not(:focus-within) {
    margin-bottom: auto !important;
    visibility: visible !important;
    transition: all 0.2s ease-in-out 0s !important;
    pointer-events: none;
    transform: translateY(-2em);
	opacity: 0;
}
```
Like I mentioned before, you could also use javascript to actually set the `hidden` attribute on receiving a blur event. That would sort of be cleaner in theory, since firefox already uses the `hidden` property to hide the findbar. But in practice it's basically the same thing as using `:not(:focus-within)`

**Reply by silverwings15** [reply]
you're spot on as always, especially the part about reflow which i encountered myself while messing around with the values. amazing!

**Comment by silverwings15** [comment]
noticed that you merged `findbarmatcheslabel.uc.js` into `findbarmods.uc.js` (along with the autohide setting, yay) so i'm playing around with it and `uc-findbar.css` now in a blank profile, pretty fun. i found this redundant comment in `uc-findbar.css`, line 71 that mentions `findbarMatchesLabel.uc.js`, just a headsup

**Reply by silverwings15** [reply]
i also found a small quirk: if i add the following CSS

`.findbar-case-sensitive {`
`display: none !important;`
`}`

the miniature matches label is no longer displayed. however, there is no issue if i hide `.findbar-entire-word` the same way

edit: my current findbar appearance :D there's one more quirk i need to figure out later about the findbar resizing itself in 'Find as you type mode' messing up the positioning of the miniature label

![image](https://user-images.githubusercontent.com/10742312/134806436-9b2c7ffd-4e3c-4904-a8ec-f2985653c6aa.png)

**Reply by aminomancer** [reply]
It's not really that I merged it into a script, I just renamed it because the number of features grew beyond the name of the script, so it felt kind of inaccurate to call it findbar matches label. Thanks for the heads up on the code comment. I recently discovered that vs code has a function for searching every file in a workspace for any text query, so that probably won't happen again.

As for the case sensitivity button, that's not really a quirk. The script decides where to position the miniature label based on where the case sensitivity button is. It's meant to appear to the left of the case sensitivity button. Its position can't be left to a relative position because it's supposed to appear as if it's like, just pigment on the background of the input field. Like if you click the label, it clicks through to the input field behind it. Otherwise there'd just be a useless click field on its bounding box.

But there isn't any relative way to do that due to the way it's positioned, between the input field and the buttons. For the input field to extend through the label, the label needs to have `position: absolute`. But that means its position is relative to the parent. So the only relative positions would be all the way to the left of the findbar or all the way to the right of the findbar. I thought about positioning it somewhere else but nowhere else really makes sense, except for outside of the textbox altogether.

So it just has absolute positioning and it gets its horizontal offset based on the distance between the left edge of the first button (case sensitive button) and the right edge of the findbar itself, the element to which the label's position is relative. That means if there's no case sensitive button, there's no way to calculate the position, hence the horizontal offset is 0. Meaning it starts at the right edge of the findbar so all of its text overflows out of the findbar's bounding box. If you use `visibility: collapse` instead of `display: none` you might get better results though.

Maybe there's a better way to do this, like use the left edge of whichever is the first visible button after the label. The problem with that is I'm not sure if any of the buttons are reliably "visible" while the findbar is initially opening. So it's really just not designed to be used with CSS that hides the case sensitive button. It could always be refactored to use some other button as the anchor instead though, i.e. `.findbar-find-previous`.

As for type ahead it's meant to be disabled. It messes up other things with my scripts, and I have had it disabled myself for a long time due to its keyboard listeners breaking various webpages. This particular issue could be solved but it would make the script less performant. The position of the label would need to be recalculated in response to a greater number of things.

I guess I forgot to mention that in the script or in the readme, but the pref is `accessibility.typeaheadfind`. It's disabled by default nowadays so I haven't thought about it in a while. I'll look into the feasibility of making the script compatible with type ahead though.

**Reply by aminomancer** [reply]
So I updated uc-findbar.css and findbarMods.uc.js. Back when I originally made this script I didn't really know what I was doing, so I had a lot of inefficient and sloppy event listeners that were modifying the findbar in response to the findbar being created. Sort of modding it from the outside in. But since then I found a much better way to do things, modifying the class that gets registered as the `<findbar>` element itself. With that I can make it so the label position only changes in response to the findbar opening, since that only happens after all the "find-as-you-type" UI changes.

I had also forgotten that the label is not even supposed to appear at all in quick find/type ahead mode. So among other fixes related to quick find, it seems I solved the problem of the label moving around (or failing to move around) in response to findbar mode shifts. So if you download the new version, you should be able to leave typeaheadfind enabled while still using the script

**Reply by silverwings15** [reply]
oh hmm, i was sure both scripts were coexisting at some point, hence i figured they got merged now that there is only one left. after learning that the case sensitivity button is necessary for the matches label to appear (the part about not making the matches label into an useless click field makes perfect sense), i just changed its color so it blends into the background instead. 

and wow, super quick work on the script and css, i'll definitely check it out as soon as i can.

**Reply by aminomancer** [reply]
Nope, it was added August 26 of this year. You can see the complete file history in vs code's gitlens or github itself.

And no I think you misunderstood my previous post. The case sensitivity button is definitely not necessary anymore. I updated the script. It uses the right edge of the textbox now, not the left edge of the case sensitivity button. Technically, it just needed to perform that calculation a bit later so that the textbox would have time to reflow in response to the case sensitivity button being hidden by quick find mode.

But all that became irrelevant since I realized the label isn't supposed to show in quick find mode anyway. Hence, it only needs to recalculate when opening in normal mode. Which effectively means it doesn't need to recalculate at all. But it still recalculates on findbar open, just in case the user sets some unusual CSS that dynamically affects the textbox's dimensions/position even in normal mode. So it should be fine whether you hide any buttons or not. Although there may still be issues if you hide the buttons dynamically under unusual conditions.

**Reply by silverwings15** [reply]
no no, i completely understood. i was simply mentioning my workaround for your 'previous' version of the findbar which i'm still currently using, as i haven't gotten around to updating it yet 😅

**Reply by aminomancer** [reply]
oh gotcha. yeah it's a lot of updates to keep up with. I wish I knew a better way of distributing this stuff. I'm sure it would be easy to create a simple installer/updater for picking and choosing the files and keeping them up to date automatically, but I don't know much about any OS scripting languages. if it were feasible I would simply write it as a firefox extension, or even an autoconfig script for installing other autoconfig scripts haha.

actually, maybe that is possible. do you think that would actually be worthwhile? it seems like, if I could make it work for my own scripts, I could probably make it function as a kind of script store in general. I know someone was doing that with CSS stylesheets recently, with a lot of hype, but seems like development just abruptly stopped a couple months ago.

**Reply by silverwings15** [reply]
hmm definitely feasible for (some of) the scripts, but wouldn't it get too complicated to make the CSS components modular to the point where they can seamlessly be plugged into any setup? especially those with preexisting CSS. plus, your repo is best used in its entirety anyway, i would too if i could figure out the hundreds of color values you used in uc-globals.css and convert those to a light theme 😄. i tried messing around with that for half an hour before giving up entirely lol

i updated the findbar stuff, and it works just fine on a new profile, but in my current profile, the label still disappears when the case sensitivity button is hidden. then i tried copying the CSS in my profile (which includes some modifications i made for compatibility) over to the new profile, and it still works on there. not sure what the issue is, but i'll enjoy getting to the bottom of this when i have time in the next day or two

**Reply by aminomancer** [reply]
Did you clear your startup cache? In what way are you hiding the button? display or visibility? Both of them are working fine for me, though it looks better with a few px margin-right when the case sensitivity button is hidden. I guess because the entire-word button has less empty space. I would try setting the CSS the way you want it and then restart firefox while clearing the startup cache. If you're using my appMenuMods script you can just right click the "restart" button in the app menu. Or shift click or ctrl click. Restarting shouldn't technically be necessary, as long as you open and close the findbar after modifying the CSS. But if you update the script, clearing the startup cache is always necessary, and sometimes doesn't work the first time.

**Reply by silverwings15** [reply]
oh duh, clearing startup cache did the trick.. thanks man, completely forgot about that 😬. i also added 6px margin-right per your suggestion, looks perfect now!

**Comment by KleinByte** [comment]
Hey I just wanted to let you know, I absolutely apprecate the work you put into this. Thank-you for also spending dozens of hours in writing up the entire readme that explains everything. Even your damn scripts you wrote have extensive descriptions in the code.

This project is absolutely a masterpiece. 

I really hope you never stop maintaining this, and I say that out of the upmost respect and not from entitlement. 
After getting everything setup, I cant see myself using anything else. 

Truly from the bottom of my heart, as much as words can express, I apprecate all the work you've done.

Also you 100% should setup donations for this project. Time is money as they say, and I'd love to repay your time with some money. 
Hopefully after Im more experienced(still in college), I could even help fix future issues that may arise.

**Reply by aminomancer** [reply]
Thanks man I appreciate it a lot. You're studying computer science then?

**Reply by aminomancer** [reply]
@KleinByte let me know if you can see the pink "Sponsor" button on the right side of the repo main page. I just followed the instructions verbatim, figured this was the simplest method

**Reply by KleinByte** [reply]
Done. Its working, got a monthly donation set up. Cheers.

**Reply by aminomancer** [reply]
Thanks! I added you to the readme.

**Comment by silverwings15** [comment]
@aminomancer so very glad to see this project still going strong almost 3 years on. much love
